kategory-docs/docs/docs/integrations/rx2/README.md

```diff
@@ -4,6 +4,146 @@ title: Rx2
+
+The largest quality of life improvement when using Observables in Kategory is the introduction of the [Monad Comprehension]({{ '/docs/patterns/monadcomprehensions' | relative_url }}). This library construct allows expressing asynchronous Observable sequences as synchronous code using binding/bind.
+
+#### Kategory Wrapper
+
+To wrap any existing Observable in its Kategory Wrapper counterpart you can use the extension function `k()`.
+
+```kotlin:ank
+import kategory.effects.*
+import io.reactivex.*
+import io.reactivex.subjects.*
+
+val obs = Observable.fromArray(1, 2, 3, 4, 5).k()
+obs
+```
+
+```kotlin:ank
+val flow = Flowable.fromArray(1, 2, 3, 4, 5).k()
+flow
+```
+
+```kotlin:ank
+val subject = PublishSubject.create().k()
+subject
+```
+
+You can return to their regular forms using the function `value()`.
+
+```kotlin:ank
+obs.value()
+```
+
+```kotlin:ank
+flow.value()
+```
+
+```kotlin:ank
+subject.value()
+```
+
+### Observable comprehensions
+
+The library provides instances of [`MonadError`]({{ '/docs/typeclasses/monaderror' | relative_url }}) and [`AsyncContext`]({{ '/docs/effects/async' | relative_url }}).
+
+[`AsyncContext`]({{ '/docs/effects/async' | relative_url }}) allows you to generify over datatypes that can run asynchronous code. You can use it with `ObservableKW` and `FlowableKW`.
+
+```kotlin
+fun <F> getSongUrlAsync(AC: AsyncContext<F> = asycContext()) =
+  AC.runAsync { getSongUrl() }
+
+val songObservable: ObservableKW<Url> = getUsersAsync().ev()
+val songFlowable: FlowableKW<Url> = getUsersAsync().ev()
+```
+
+[`MonadError`]({{ '/docs/typeclasses/monaderror' | relative_url }}) can be used to start a [Monad Comprehension]({{ '/docs/patterns/monadcomprehensions' | relative_url }}) using the method `bindingE`, with all its benefits.
+
+Let's take an example and convert it to a comprehension. We'll create an observable that loads a song from a remote location, and then reports the current play % every 100 milliseconds until the percentage reaches 100%:
+
+```kotlin
+getSongUrlAsync()
+  .map { songUrl -> MediaPlayer.load(songUrl) }
+  .flatMap {
+    val totalTime = musicPlayer.getTotaltime()
+    Observable.interval(100, Milliseconds)
+      .flatMap { 
+        Observable.create { musicPlayer.getCurrentTime() }
+          .subscribeOn(AndroidSchedulers.mainThread())
+          .map { tick -> (tick / totalTime * 100).toInt() }
+      }
+      .takeUntil { percent -> percent >= 100 }
+      .observeOn(Schedulers.immediate())
+  }
+```
+
+When rewritten using `bindingE` it becomes. Note that any unexpected exception, like AritmeticException when totalTime is 0, is automatically caught and wrapped inside the observable. 
+
+```kotlin
+ObservableKW.errorMonad().bindingE {
+  val songUrl = getSongUrlAsync().bind()
+  val musicPlayer = MediaPlayer.load(songUrl)
+  val totalTime = musicPlayer.getTotaltime()
+
+  val end = PublishSubject.create<Unit>()
+  Observable.interval(100, Milliseconds).takeUntil(end).bind()
```

Comment:
What would this line do exactly, somehow hold the comprehension until it finishes?

Reply:
To avoid the interval Observable running indefinitely you have to cancel it at some point, which can be regulated imperatively by using an `end` method call. The way of doing it in rx is using a local takeUntil subject.

Reply:
Yes, the takeUntil part I get, I meant why do you need to declare an observable that doesn't seem to do anything (at least you don't store the bind() result anywhere).